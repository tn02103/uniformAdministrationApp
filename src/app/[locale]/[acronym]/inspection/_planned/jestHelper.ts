import dayjs from "@/lib/dayjs";
import { PlannedInspectionType } from "@/types/inspectionTypes";


export const mockCadetList = [
    { id: "1", firstname: "Jack", lastname: "Doe" },
    { id: "2", firstname: "Jane", lastname: "Smith" },
    { id: "3", firstname: "Alice", lastname: "Johnson" },
    { id: "4", firstname: "Bob", lastname: "Brown" },
    { id: "5", firstname: "Charlie", lastname: "Davis" },
];

export const mockInspectionList: PlannedInspectionType[] = [
    {
        id: "1777e18f-f4fa-4ec4-90d8-58a9b1ce16ac",
        name: "Inspection 1",
        date: dayjs().toDate(),
        timeStart: null,
        timeEnd: null,
        deregistrations: [
            {
                fk_cadet: mockCadetList[0].id,
                cadet: mockCadetList[0],
                date: dayjs().toDate(),
                fk_inspection: "1777e18f-f4fa-4ec4-90d8-58a9b1ce16ac",
            },
            {
                fk_cadet: mockCadetList[1].id,
                cadet: mockCadetList[1],
                date: dayjs().subtract(10, "day").toDate(),
                fk_inspection: "1777e18f-f4fa-4ec4-90d8-58a9b1ce16ac",
            }
        ]
    },
    {
        id: "1777e18f-f4fa-4ec4-90d8-58a9b1ce16ad",
        name: "Inspection 2",
        date: dayjs().add(2, "day").toDate(),
        timeStart: null,
        timeEnd: null,
        deregistrations: [
            {
                fk_cadet: mockCadetList[2].id,
                cadet: mockCadetList[2],
                date: dayjs().toDate(),
                fk_inspection: "1777e18f-f4fa-4ec4-90d8-58a9b1ce16ad",
            }
        ]
    },
    {
        id: "1777e18f-f4fa-4ec4-90d8-58a9b1ce16ae",
        name: "Inspection 3",
        date: new Date("2023-10-01"),
        timeStart: new Date("2023-10-01T10:00:00"),
        timeEnd: null,
        deregistrations: []
    }
];

jest.mock("@/dal/inspection", () => ({
    createInspection: jest.fn().mockResolvedValue("createInspection"),
    deleteInspection: jest.fn().mockResolvedValue("deleteInspection"),
    updatePlannedInspection: jest.fn().mockResolvedValue("updateInspection"),
    updateCadetRegistrationForInspection: jest.fn().mockResolvedValue("updateCadetRegistrationForInspection"),

    startInspection: jest.fn().mockResolvedValue("startInspection"),
    stopInspection: jest.fn().mockResolvedValue("stopInspection"),
}));
jest.mock("@/dataFetcher/inspection", () => ({
    usePlannedInspectionList: jest.fn().mockReturnValue({
        data: mockInspectionList,
        mutate: jest.fn().mockResolvedValue(mockInspectionList),
    }),
    useInspectionState: jest.fn().mockReturnValue({
        data: { active: false },
        mutate: jest.fn().mockResolvedValue({ active: false }),
    }),
}))
