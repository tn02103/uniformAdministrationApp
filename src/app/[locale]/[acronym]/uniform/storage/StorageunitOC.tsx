import { LabelIconButton } from "@/components/Buttons/LabelIconButton";
import { useModal } from "@/components/modals/modalProvider";
import { deleteStorageUnit, StorageUnitWithUniformItems } from "@/dal/storageUnit/_index";
import { useStorageUnitsWithUniformItemList } from "@/dataFetcher/storage";
import { useI18n } from "@/lib/locales/client";
import { useState } from "react";
import { Offcanvas, Row } from "react-bootstrap";
import { toast } from "react-toastify";
import {StorageunitOCDetailForm} from "./StorageunitOCDetailForm";
import { StorageunitOCHeader } from "./StorageunitOCHeader";
import StorageunitOCUniformList from "./StorageunitOCUniformList";

type Props = {
    storageUnit?: StorageUnitWithUniformItems;
    onHide: () => void;
    setSelectedStorageUnitId: (id: string | null) => void;
}
export default function StorageunitOC({ storageUnit, onHide, setSelectedStorageUnitId }: Props) {
    const [editable, setEditable] = useState(!storageUnit);
    const { mutate } = useStorageUnitsWithUniformItemList();
    const t = useI18n();
    const modal = useModal();

    const handleOnHide = () => {
        if (editable) {
            modal?.showMessageModal(
                'Änderungen verwerfen?',
                'Möchten Sie die Änderungen verwerfen?',
                [
                    { option: 'Abbrechen', function: () => { }, type: 'secondary', testId: 'btn_no' },
                    { option: 'Verwerfen', function: onHide, type: 'danger', testId: 'btn_yes' },
                ],
                'warning',
            );
        } else {
            onHide();
        }
    }

    const handleDelete = () => {
        if (!storageUnit) return;

        const deleteMutation = () => deleteStorageUnit(storageUnit.id).then((data) => {
            mutate(data);
            onHide();
        }).catch(() => {
            toast.error(t('common.error.actions.delete'));
        });

        modal?.showMessageModal(
            'Löschen',
            'Möchten Sie den Eintrag wirklich löschen?',
            [
                { option: 'Abbrechen', function: () => { }, type: 'secondary', testId: 'btn_no' },
                { option: 'Löschen', function: deleteMutation, type: 'danger', testId: 'btn_yes' },
            ],
            'danger'
        );
    }


    return (
        <Offcanvas show={true} onHide={handleOnHide} placement='end' style={{ width: '500px' }}>
            <StorageunitOCHeader storageUnit={storageUnit} />
            <Offcanvas.Body>
                <hr className="mb-0" />
                <Row className="mb-3 justify-content-evenly">
                    <LabelIconButton
                        variantKey="edit"
                        disabled={editable || !storageUnit}
                        onClick={() => setEditable(true)} />
                    <LabelIconButton
                        variantKey="delete"
                        onClick={handleDelete}
                        disabled={!storageUnit || editable}
                    />
                </Row>
                <StorageunitOCDetailForm
                    editable={editable}
                    storageUnit={storageUnit}
                    setEditable={setEditable}
                    setSelectedStorageUnitId={setSelectedStorageUnitId}
                    onHide={onHide}
                />
                {storageUnit && (
                    <>
                        <hr className="mb-0" />
                        <StorageunitOCUniformList storageUnit={storageUnit} />
                    </>
                )}
            </Offcanvas.Body>
        </Offcanvas>
    );
}