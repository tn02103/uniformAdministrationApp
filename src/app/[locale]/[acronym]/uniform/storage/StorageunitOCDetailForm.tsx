import type { StorageUnitWithUniformItems } from "@/dal/storageUnit/_index";

import { Form } from "@/components/fields/Form";
import { InputFormField } from "@/components/fields/InputFormField";
import { NumberInputFormField } from "@/components/fields/NumberInputFormField";
import { TextareaFormField } from "@/components/fields/TextareaFormField";
import { ToggleFormField } from "@/components/fields/ToggleFormField";
import { createStorageUnit, updateStorageUnit } from "@/dal/storageUnit/_index";
import { useStorageUnitsWithUniformItemList } from "@/dataFetcher/storage";
import { useI18n, useScopedI18n } from "@/lib/locales/client";
import { SAFormHandler } from "@/lib/SAFormHandler";
import { storageUnitFormSchema, StorageUnitFormType } from "@/zod/storage";
import { Button, Col, Row } from "react-bootstrap";
import { useFormContext } from "react-hook-form";
import z from "zod";


type Props = {
    editable?: boolean;
    storageUnit?: StorageUnitWithUniformItems;
    setEditable: (editable: boolean) => void;
    setSelectedStorageUnitId: (id: string | null) => void;
    onHide: () => void;
}

export function StorageunitOCDetailForm({ editable, storageUnit, setEditable, setSelectedStorageUnitId, onHide }: Props) {
    const t = useI18n();
    const { mutate, storageUnits } = useStorageUnitsWithUniformItemList();

    let formSchema;
    if (!storageUnit) {
        formSchema = storageUnitFormSchema.refine((schema) => storageUnits?.every(s => s.name !== schema.name), {
            message: t('storageUnit.error.nameDuplication'),
            path: ['name'],
        });
    } else {
        formSchema = storageUnitFormSchema.omit({ name: storageUnit ? true : undefined })
    }
    type FormType = z.infer<typeof formSchema>;

    let defaultValue: FormType | undefined = undefined;
    if (storageUnit) {
        const zodData = formSchema.safeParse(storageUnit);
        if (zodData.success) {
            defaultValue = zodData.data;
        }
    }

    const handleSave = async (data: FormType) => {
        if (!storageUnit) {
            await SAFormHandler(
                createStorageUnit(data as StorageUnitFormType),
                null,
                (returnData) => {
                    if (Array.isArray(returnData)) {
                        mutate(returnData);
                        setEditable(false);
                        setSelectedStorageUnitId(returnData.find(u => u.name === (data as StorageUnitFormType).name)?.id ?? null);
                    }
                },
                t('common.error.actions.create'),
            );
        }
        else {
            await SAFormHandler(
                updateStorageUnit({ id: storageUnit.id, data }),
                null,
                (returnData) => {
                    mutate(returnData);
                    setEditable(false);
                },
                t('common.error.actions.save'),
            );
        }
    }
    const handleCancel = () => {
        if (storageUnit) {
            setEditable(false);
        } else {
            onHide();
        }
    }
    return (
        <Form<FormType>
            onSubmit={handleSave}
            formName="storageUnitForm"
            disabled={!editable}
            plaintext={!editable}
            defaultValues={defaultValue}
            zodSchema={formSchema}
        >
            <Row>
                {!storageUnit && (
                    <Col xs={12} className="mb-3">
                        <InputFormField name="name" label={t('storageUnit.label.details.name')} required />
                    </Col>
                )}
                <Col xs={12} className="mb-3">
                    <TextareaFormField name="description" label={t('storageUnit.label.details.description')} />
                </Col>
                <Col xs={4} className="mb-3">
                    <NumberInputFormField name="capacity" label={t('storageUnit.label.details.capacity')} className="" />
                </Col>
                <Col xs={7} className="mb-3">
                    <ToggleFormField
                        name="isReserve"
                        label={t('storageUnit.label.details.forReserves')}
                        toggleText={t('storageUnit.label.details.forReservesText')}
                    />
                </Col>
            </Row>
            {editable && (
                <Row>
                    <Col>
                        <CancelButton onClick={handleCancel} />
                    </Col>
                    <Col>
                        <Button type="submit" size="sm" variant="outline-primary" className="m-1">
                            {t('common.actions.save')}
                        </Button>
                    </Col>
                </Row>
            )}
        </Form>
    );
}

const CancelButton = ({ onClick }: { onClick: () => void }) => {
    const { reset } = useFormContext();
    const t = useScopedI18n('common.actions');

    const handleCancel = () => {
        reset();
        onClick();
    }

    return (
        <Button type="button" size="sm" variant="outline-secondary" className="m-1" onClick={handleCancel}>
            {t('cancel')}
        </Button>
    )
};
