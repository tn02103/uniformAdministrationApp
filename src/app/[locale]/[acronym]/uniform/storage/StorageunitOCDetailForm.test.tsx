import "./StorageunitOC.jestHelper";

import { getByRole, queryByRole, render } from "@testing-library/react";
import { StorageunitOCDetailForm } from "./StorageunitOCDetailForm";
import userEvent from "@testing-library/user-event";
import { mockStorageUnitWithItems } from "./StorageunitOC.jestHelper";



describe("StorageunitOCDetailForm", () => {
    const { toast } = jest.requireMock("react-toastify");

    const setEditable = jest.fn();
    const setSelectedStorageUnitId = jest.fn();
    const onHide = jest.fn();

    beforeEach(() => {
        jest.clearAllMocks();
    });

    it("should render without crashing", () => {
        const { container } = render(
            <StorageunitOCDetailForm
                editable={true}
                storageUnit={undefined}
                setEditable={jest.fn()}
                setSelectedStorageUnitId={jest.fn()}
                onHide={jest.fn()}
            />
        );

        expect(container).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    describe("editable state", () => {
        it("renders editable state", () => {
            const { container } = render(
                <StorageunitOCDetailForm
                    editable={true}
                    storageUnit={mockStorageUnitWithItems[0]}
                    setEditable={setEditable}
                    setSelectedStorageUnitId={setSelectedStorageUnitId}
                    onHide={onHide}
                />
            );

            const nameInput = queryByRole(container, "textbox", { name: /name/i });
            const capacityInput = getByRole(container, "spinbutton", { name: /capacity/i });
            const descriptionInput = getByRole(container, "textbox", { name: /description/i });
            const reserveSwitch = getByRole(container, "switch", { name: /reserve/i });

            expect(nameInput).toBeNull();
            expect(capacityInput).toBeInTheDocument();
            expect(descriptionInput).toBeInTheDocument();
            expect(reserveSwitch).toBeInTheDocument();

            expect(capacityInput).toBeEnabled();
            expect(descriptionInput).toBeEnabled();
            expect(reserveSwitch).toBeEnabled();
        });
        it("renders non-editable state", () => {
            const { container } = render(
                <StorageunitOCDetailForm
                    editable={false}
                    storageUnit={mockStorageUnitWithItems[0]}
                    setEditable={setEditable}
                    setSelectedStorageUnitId={setSelectedStorageUnitId}
                    onHide={onHide}
                />
            );

            const nameInput = queryByRole(container, "textbox", { name: /name/i });
            const capacityInput = getByRole(container, "spinbutton", { name: /capacity/i });
            const descriptionInput = getByRole(container, "textbox", { name: /description/i });
            const reserveSwitch = getByRole(container, "switch", { name: /reserve/i });

            expect(nameInput).not.toBeInTheDocument();
            expect(capacityInput).toBeInTheDocument();
            expect(descriptionInput).toBeInTheDocument();
            expect(reserveSwitch).toBeInTheDocument();

            expect(capacityInput).toBeDisabled();
            expect(descriptionInput).toBeDisabled();
            expect(reserveSwitch).toHaveAttribute("aria-disabled", "true");
        });
    });

    describe("with storage unit", () => {
        const mockData = {
            id: "19518cad-dd54-44fb-9e31-29ba4ed9b7f9",
            name: "Updated Box 1",
            description: "This is an updated test storage unit.",
            capacity: 15,
            uniformList: [],
        }
        it("should update existing storage unit when unit is provided", async () => {
            const { updateStorageUnit } = jest.requireMock("@/dal/storageUnit/_index");
            updateStorageUnit.mockResolvedValue(mockStorageUnitWithItems.map(item => item.id === mockData.id ? mockData : item));

            const user = userEvent.setup();
            const { container } = render(
                <StorageunitOCDetailForm
                    editable={true}
                    storageUnit={mockStorageUnitWithItems[0]}
                    setEditable={setEditable}
                    setSelectedStorageUnitId={setSelectedStorageUnitId}
                    onHide={onHide}
                />
            );

            expect(queryByRole(container, "textbox", { name: /name/i })).toBeNull();

            const capacityInput = getByRole(container, "spinbutton", { name: /capacity/i });
            const descriptionInput = getByRole(container, "textbox", { name: /description/i });
            const reserveSwitch = getByRole(container, "switch", { name: /reserve/i });

            await user.clear(capacityInput);
            await user.clear(descriptionInput);
            await user.type(capacityInput, mockData.capacity.toString());
            await user.type(descriptionInput, mockData.description);
            await user.click(reserveSwitch);

            await user.click(getByRole(container, "button", { name: /save/i }));

            expect(updateStorageUnit).toHaveBeenCalledWith({
                id: mockData.id,
                data: {
                    description: mockData.description,
                    capacity: mockData.capacity,
                    isReserve: !mockStorageUnitWithItems[0].isReserve,
                },
            });
            expect(setEditable).toHaveBeenCalledWith(false);
            expect(onHide).not.toHaveBeenCalled();
            expect(setSelectedStorageUnitId).not.toHaveBeenCalled();
        });

        it("should cancel editing when cancel button is clicked", async () => {
            const user = userEvent.setup();
            const { container } = render(
                <StorageunitOCDetailForm
                    editable={true}
                    storageUnit={mockStorageUnitWithItems[0]}
                    setEditable={setEditable}
                    setSelectedStorageUnitId={setSelectedStorageUnitId}
                    onHide={onHide}
                />
            );
            const capacityInput = getByRole(container, "spinbutton", { name: /capacity/i });
            const descriptionInput = getByRole(container, "textbox", { name: /description/i });

            await user.clear(capacityInput);
            await user.clear(descriptionInput);
            await user.type(capacityInput, mockData.capacity.toString());
            await user.type(descriptionInput, mockData.description);

            await user.click(getByRole(container, "button", { name: /cancel/i }));

            expect(capacityInput).toHaveValue(mockStorageUnitWithItems[0].capacity);
            expect(descriptionInput).toHaveValue(mockStorageUnitWithItems[0].description);
            expect(setEditable).toHaveBeenCalledWith(false);
            expect(onHide).not.toHaveBeenCalled();
            expect(setSelectedStorageUnitId).not.toHaveBeenCalled();
        });

        it("should catch DAL exceptions on update", async () => {
            const { updateStorageUnit } = jest.requireMock("@/dal/storageUnit/_index");
            updateStorageUnit.mockRejectedValue(new Error("Test error"));

            const user = userEvent.setup();
            const { container } = render(
                <StorageunitOCDetailForm
                    editable={true}
                    storageUnit={mockStorageUnitWithItems[0]}
                    setEditable={setEditable}
                    setSelectedStorageUnitId={setSelectedStorageUnitId}
                    onHide={onHide}
                />
            );

            await user.click(getByRole(container, "button", { name: /save/i }));

            expect(updateStorageUnit).toHaveBeenCalled();
            expect(toast.error).toHaveBeenCalled();

            expect(setEditable).not.toHaveBeenCalled();
            expect(onHide).not.toHaveBeenCalled();
        });
    });

    describe("without storage unit", () => {
        const mockData = {
            id: "new-id",
            name: "New Storage Unit",
            description: "This is a test storage unit.",
            capacity: 20,
            uniformList: [],
        }
        it("should create new storage unit", async () => {
            const { createStorageUnit } = jest.requireMock("@/dal/storageUnit/_index");
            createStorageUnit.mockResolvedValue([...mockStorageUnitWithItems, mockData]);

            const user = userEvent.setup();
            const { container } = render(
                <StorageunitOCDetailForm
                    editable={true}
                    storageUnit={undefined}
                    setEditable={setEditable}
                    setSelectedStorageUnitId={setSelectedStorageUnitId}
                    onHide={onHide}
                />
            );

            await user.type(getByRole(container, "textbox", { name: /name/i }), mockData.name);
            await user.type(getByRole(container, "spinbutton", { name: /capacity/i }), mockData.capacity.toString());
            await user.type(getByRole(container, "textbox", { name: /description/i }), mockData.description);

            await user.click(getByRole(container, "button", { name: /save/i }));

            expect(createStorageUnit).toHaveBeenCalledWith({
                name: mockData.name,
                isReserve: false,
                description: mockData.description,
                capacity: mockData.capacity,
            });
            expect(setEditable).toHaveBeenCalledWith(false);
            expect(setSelectedStorageUnitId).toHaveBeenCalledWith(mockData.id);
            expect(onHide).not.toHaveBeenCalled();
        });

        it("should catch name duplication error on create", async () => {
            const { createStorageUnit } = jest.requireMock("@/dal/storageUnit/_index");
            createStorageUnit.mockResolvedValue({
                error: {
                    formElement: "name",
                    message: "custom.nameDuplication.storageUnit",
                },
            });

            const user = userEvent.setup();
            const { container } = render(
                <StorageunitOCDetailForm
                    editable={true}
                    storageUnit={undefined}
                    setEditable={setEditable}
                    setSelectedStorageUnitId={setSelectedStorageUnitId}
                    onHide={onHide}
                />
            );

            await user.type(getByRole(container, "textbox", { name: /name/i }), mockStorageUnitWithItems[0].name);
            await user.type(getByRole(container, "spinbutton", { name: /capacity/i }), mockData.capacity.toString());
            await user.type(getByRole(container, "textbox", { name: /description/i }), mockData.description);

            await user.click(getByRole(container, "button", { name: /save/i }));
            expect(createStorageUnit).not.toHaveBeenCalled();

            expect(getByRole(container, "alert", { name: /name/i })).toHaveTextContent(/nameDuplication/i);
        });

        it("should call onHide if cancel with no storage unit", async () => {
            const user = userEvent.setup();
            const { container } = render(
                <StorageunitOCDetailForm
                    editable={true}
                    storageUnit={undefined}
                    setEditable={setEditable}
                    setSelectedStorageUnitId={setSelectedStorageUnitId}
                    onHide={onHide}
                />
            );

            await user.click(getByRole(container, "button", { name: /cancel/i }));

            expect(onHide).toHaveBeenCalled();
            expect(setEditable).not.toHaveBeenCalled();
            expect(setSelectedStorageUnitId).not.toHaveBeenCalled();
        });

        it("should catch DAL exceptions on create", async () => {
            const { createStorageUnit } = jest.requireMock("@/dal/storageUnit/_index");
            createStorageUnit.mockRejectedValue(new Error("Test error"));

            const user = userEvent.setup();
            const { container } = render(
                <StorageunitOCDetailForm
                    editable={true}
                    storageUnit={undefined}
                    setEditable={setEditable}
                    setSelectedStorageUnitId={setSelectedStorageUnitId}
                    onHide={onHide}
                />
            );

            await user.type(getByRole(container, "textbox", { name: /name/i }), mockData.name);
            await user.type(getByRole(container, "spinbutton", { name: /capacity/i }), mockData.capacity.toString());
            await user.type(getByRole(container, "textbox", { name: /description/i }), mockData.description);

            await user.click(getByRole(container, "button", { name: /save/i }));

            expect(createStorageUnit).toHaveBeenCalled();
            expect(toast.error).toHaveBeenCalled();

            expect(setEditable).not.toHaveBeenCalled();
            expect(onHide).not.toHaveBeenCalled();
        });
    });
});
