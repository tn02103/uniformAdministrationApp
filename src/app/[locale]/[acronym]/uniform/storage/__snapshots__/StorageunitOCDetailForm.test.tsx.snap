// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StorageunitOCDetailForm should render without crashing 1`] = `
<div>
  <form
    autocomplete="off"
    class="form"
    novalidate=""
  >
    <div
      class="row"
    >
      <div
        class="mb-3 col-12"
      >
        <div>
          <label
            class="fw-bold m-0 undefined form-label"
            for="undefined_input-name"
          >
            storageUnit.label.details.name
             *
          </label>
          <input
            aria-invalid="false"
            aria-required="true"
            class="form-control"
            id="undefined_input-name"
            name="name"
            type="text"
            value=""
            width="auto"
          />
          <div
            aria-label="error message name"
            class="text-danger fs-7"
            data-testid="err_name"
            id="undefined_err_name"
            role="alert"
          />
        </div>
      </div>
      <div
        class="mb-3 col-12"
      >
        <div>
          <label
            class="fw-bold m-0 undefined form-label"
            for="storageUnitForm_input-description"
          >
            storageUnit.label.details.description
          </label>
          <textarea
            aria-invalid="false"
            class="form-control"
            id="storageUnitForm_input-description"
            name="description"
            rows="3"
            width="auto"
          />
          <div
            aria-label="error message description"
            class="text-danger fs-7"
            data-testid="err_description"
            id="storageUnitForm_err_description"
            role="alert"
          />
        </div>
      </div>
      <div
        class="mb-3 col-4"
      >
        <div>
          <label
            class="fw-bold m-0 undefined form-label"
            for="storageUnitForm_numberInput-capacity"
          >
            storageUnit.label.details.capacity
          </label>
          <input
            aria-invalid="true"
            class="form-control"
            id="storageUnitForm_numberInput-capacity"
            name="capacity"
            type="number"
            value=""
          />
          <div
            aria-label="error message capacity"
            class="text-danger fs-7"
            data-testid="err_capacity"
            id="storageUnitForm_err_capacity"
            role="alert"
          />
        </div>
      </div>
      <div
        class="mb-3 col-7"
      >
        <div>
          <label
            class="fw-bold m-0 undefined form-label"
            for="storageUnitForm_toggle-isReserve"
          >
            storageUnit.label.details.forReserves
          </label>
          <div
            class="form-check form-switch px-0 my-auto"
          >
            <input
              aria-checked="false"
              aria-describedby="storageUnitForm_isReserve_toggleText"
              class="form-check-input ms-0 my-2"
              id="storageUnitForm_toggle-isReserve"
              name="isReserve"
              role="switch"
              style="cursor: pointer;"
              tabindex="0"
              type="checkbox"
              value="false"
            />
            <label
              class="form-check-label ms-2"
              id="storageUnitForm_isReserve_toggleText"
            >
              storageUnit.label.details.forReservesText
            </label>
          </div>
          <div
            aria-label="error message isReserve"
            class="text-danger fs-7"
            data-testid="err_isReserve"
            id="storageUnitForm_err_isReserve"
            role="alert"
          />
        </div>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col"
      >
        <button
          class="m-1 btn btn-outline-secondary btn-sm"
          type="button"
        >
          common.actions.cancel
        </button>
      </div>
      <div
        class="col"
      >
        <button
          class="m-1 btn btn-outline-primary btn-sm"
          type="submit"
        >
          common.actions.save
        </button>
      </div>
    </div>
  </form>
</div>
`;
